src/Xamarin.Android.Build.Tasks/Utilities/NativeAssemblyGenerator/NativeAssemblyGenerator.cs

```diff
@@ -0,0 +1,1223 @@
+		{
+			return arch switch {
+				AndroidTargetArch.Arm => new Arm32NativeAssemblyGenerator (output, fileName),
+				AndroidTargetArch.Arm64 => new Arm64NativeAssemblyGenerator (output, fileName),
+				AndroidTargetArch.X86 => new X86_32NativeAssemblyGenerator (output, fileName),
+				AndroidTargetArch.X86_64 => new X86_64NativeAssemblyGenerator (output, fileName),
+				_ => throw new InvalidOperationException ($"Unsupported Android target ABI {arch}")
+			};
+		}
+
+		public static string GetAbiName (AndroidTargetArch arch)
+		{
+			return arch switch {
+				AndroidTargetArch.Arm => "armeabi-v7a",
+				AndroidTargetArch.Arm64 => "arm64-v8a",
+				AndroidTargetArch.X86 => "x86",
+				AndroidTargetArch.X86_64 => "x86_64",
+				_ => throw new InvalidOperationException ($"Unsupported Android architecture: {arch}"),
+			};
+		}
+
+		protected void ConfigureTypeMapping<T> (string name, uint size, uint alignment)
+		{
+			typeMapping[typeof(T)] = new NativeType {
+				Size = size,
+				Alignment = alignment,
+				Name = name,
+			};
+		}
+
+		protected abstract NativeType GetPointerType ();
+
+		protected virtual void ConfigureTypeMappings (Dictionary<Type, NativeType?> mapping)
+		{
+			ConfigureTypeMapping<byte> (".byte", size: 1, alignment: 1);
+			ConfigureTypeMapping<bool> (".byte", size: 1, alignment: 1);
+			ConfigureTypeMapping<sbyte> (".byte", size: 1, alignment: 1);
+		}
+
+		void EnsureTypeMapping ()
+		{
+			if (typeMappingConfigured) {
+				return;
+			}
+
+			ConfigureTypeMappings (typeMapping);
+
+			foreach (var kvp in typeMapping) {
+				Type managedType = kvp.Key;
+				NativeType? nativeType = kvp.Value;
+
+				if (nativeType == null) {
+					throw new InvalidOperationException ($"Missing managed nativeType {managedType} mapping");
+				}
+
+				if (nativeType.Size == 0) {
+					throw new InvalidOperationException ($"Missing size of native nativeType corresponding to managed nativeType {managedType}");
+				}
+
+				if (nativeType.Alignment == 0) {
+					throw new InvalidOperationException ($"Missing byte alignment of native nativeType corresponding to managed nativeType {managedType}");
+				}
+
+				if (String.IsNullOrEmpty (nativeType.Name)) {
+					throw new InvalidOperationException ($"Missing name of native nativeType corresponding to managed nativeType {managedType}");
+				}
+			}
+
+			typeMappingConfigured = true;
+		}
+
+		public virtual void WriteFileTop ()
+		{
+			WriteDirective (".file", QuoteString (fileName));
+		}
+
+		public virtual void WriteFileEnd ()
+		{
+			Output.WriteLine ();
+
+			if (stringTable.Count > 0) {
+				WriteStringSection (Output, AutomaticStringSection);
+
+				foreach (LabeledStringSymbol sym in stringTable) {
+					WriteSymbol (Output, sym, local: true, ownSection: false);
+				}
+
+				Output.WriteLine ();
+			}
+
+			WriteDirective (".ident", QuoteString ($"Xamarin.Android {XABuildConfig.XamarinAndroidVersion}"));
```

Comment:
Why do we need to start emitting `ident "Xamarin.Android 12.2.99"`?

What do we start doing when .NET 7+ is our primary distribution vehicle and not Classic Xamarin.Android? Presumably we'll stop bumping or otherwise remove `$(ProductVersion)` *eventually*, and then use `$(AndroidPackVersion)` instead.  (Presumably?). Should use use `$(ProductVersion)` for both .NET 6+ & Classic now, `$(AndroidPackVersion)` for both now, or selectively use one depending on which is being used?

Reply:
I don't really care which version we use, I wanted to have `Xamarin.Android ${VERSION}` in there, it might be helpful in diagnosing issues to know which XA built the binary. We can put git commit hash there, whatever you think makes the most sense.

Reply:
I kinda like the `xamarin/xamarin-android@COMMIT-HASH` idea, actually… Especially since it's "product name agnostic", and thus avoids the entire question in the first place.

Reply:
I think it still should have some "product" name, not just the hash. `Xamarin.Android [HASH]`?

Reply:
A "product name" of `xamarin/xamarin-android@` avoids the need (responsibility) to pick a name, and since I'm being incredibly wishy-washy here, abdicating that responsibility sounds awesome.

…or we just stick with `Xamarin.Android`. :-).

The hash remains useful, and should be extremely helpful for future issue investigations.